README
You can find the latest original documentation at
http://gentoo.neysx.org/mystuff/gorg/gorg.xml



How to install Gorg

Xavier Neys  Author

1. Introduction

Gorg allows you to serve your own local copy of http://www.gentoo.org. It can
use either a cgi or a fastcgi script with apache, or even use its own
stand-alone web server. Its name is short for Gentoo.org.

Of course, it can also be used in any other environment to apply XSL to some XML.

Gorg has been tested with the following packages:

Code Listing 1.1: Test environment

[ebuild   R   ] net-www/apache-2.0.53
[ebuild   R   ] net-www/mod_fcgid-1.05
[ebuild   R   ] dev-lang/ruby-1.8.2
[ebuild   R   ] dev-libs/fcgi-2.4.0
[ebuild   R   ] dev-ruby/ruby-fcgi-0.8.5-r1
[ebuild   R   ] dev-libs/libxml2-2.6.18
[ebuild   R   ] dev-libs/libxslt-1.1.13

Note: At the time of writing, mod_fcgid was not in Portage yet. Please see
http://bugs.gentoo.org/show_bug.cgi?id=79313 for an ebuild

2. Installing Gorg

Download the gorg ebuild and drop it into your Portage overlay.

Compute its digest with ebuild gorg-0.5.ebuild digest. This command will
download the tarball automatically. Finally, run emerge gorg.

Gorg uses the apache2 and fastcgi USE flags.

3. Configuring Gorg

Configuring apache 

You may skip this section if you are not going to use apache at all.

If you want to use fastcgi, which you should anyway, you'll need to add -D
FCGID to the APACHE2_OPTS variable in /etc/conf.d/apache2.

Then, integrate the apache configuration directives from the provided sample
vhost configuration file /etc/gorg/vhost.sample into your own apache configuration.
Comments in the sample config file will guide you.

Configuring Gorg 

Create a copy of the sample config file /etc/gorg/gorg.conf.sample named /etc/
gorg/gorg.conf and edit it. Comments will help you define your own parameters.

If you do not want to use the default /etc/gorg/gorg.conf config file, you'll
need to define an environment variable called GORG_CONF that points to the
config file.

Getting the missing files 

Assuming you'll serve your local copy of CVS, or a copy if, or symlinks to it,
you need to download some files from the dyn directory.

Code Listing 3.1: Get the missing files

# cd to your htdocs directory
$ cd /path/to/your/document/root
/htdocs $ mkdir -p dyn/icons
/htdocs $ cd dyn
/htdocs $ wget -O news-index.xml http://www.gentoo.org/dyn/news-index.xml?passthru=1
# If you also want the icon pages
/htdocs $ wget -O icons.xml http://www.gentoo.org/dyn/icons.xml?passthru=1
/htdocs $ for FFF in `grep '\.xml"' icons.xml|sed 's:.*\([A-Z2]\{3\}\.xml\).*:\1:'`;do wget -O icons/$FFF http://www.gentoo.org/dyn/icons/$FFF?passthru=1;done
/htdocs $ cd ..
# If you ever need other data from the dyn pages,
# I am sure you get the picture on how to do it.


You also need to make the pictures available to your browser. The images
directory is one level above htdocs. Just define a symlink to it and you're
set.

Code Listing 3.2: Make a symlink to the pictures

/htdocs $ ln -si ../images images
# It should look like this:
/htdocs $ $ ls -l
drwxr-xr-x   3 neysx users  128 Sep 14 17:45 css
drwxr-xr-x  31 neysx users  744 Oct 26 00:03 doc
drwxr-xr-x   3 neysx users  544 Nov  2 16:53 dtd
drwxr-xr-x   3 neysx users  168 Nov  3 16:24 dyn
-rw-r--r--   1 neysx users 1406 Jun  7  2003 favicon.ico
lrwxrwxrwx   1 neysx users   10 Oct 21 22:29 images -> ../images/
-rw-r--r--   1 neysx users  190 Nov  9  2002 index.xml
drwxr-xr-x  16 neysx users  384 Apr  1  2004 main
drwxr-xr-x  17 neysx users 6960 Nov  3 15:34 news
drwxr-xr-x   8 neysx users  192 Oct 23 14:52 proj
drwxr-xr-x   4 neysx users   96 Sep 17 14:05 security
drwxr-xr-x   3 neysx users  736 Nov  2 16:40 xsl

# Your local CVS probably shows a few more entries, but at least
# those mentioned above should be available and kept up-to-date.
# Also remember to keep your images directory current.


4. Running Gorg

The stand-alone web server 

The easiest way to try it out is to run gorg. It should display something like:

Code Listing 4.1: Run Gorg

$ gorg


Starting the Gorg web server on port 8008

Hit Ctrl-C or type "kill 31479" to stop it


Point your browser to http://localhost:8008 and you should see your favourite
site.

With apache 

Restart apache (/etc/init.d/apache2 restart) and visit http://localhost
assuming you're installing on your own workstation.

If it doesn't work, try the stand-alone web server (type gorg). If this doesn't
work either, check your /etc/gorg/gorg.conf config file. If it does work,
please check your apache config files and your logs. Also check that the cache directory
defined in your gorg config file exists and is writable to your apache user.
If all fails, contact me.



The contents of this document are licensed under the Creative Commons -
Attribution / Share Alike license.


Changelog
2004-12-06 gorg-0.3 initiated

2004-12-07 http://bugs.gentoo.org/show_bug.cgi?id=73163 from Masatomo Nakano <nakano@gentoo.org>
           Use PATH_TRANSLATED to resolve files in ~username

2004-12-16 Issue Etag and Last-Modified response headers fields
           Handle If-Modified-Since and If-None-Match request header fields
           Client caching will work in a consistent manner across several web nodes
           Set processed document mtime to latest mtime of dependent files and use it for Last-Modified
           Use timeouts when accessing cache (clean & store operations)
           Fork & Detach when cleaning up the cache
           Fix bugs:
             . in XSLT C extension when no document root is used
             . empty request params prevent use of cache

======================================================
           
2005-01-10 gorg-0.4 initiated

2005-01-10  Update Caching:
              . Use a dir tree under cache dir that is identical
                to the server tree. Some installations seem
                to have trouble with 30000+ files in one directory
                New param cacheTree = 1 activates this behaviour
              . Define MaxFiles param to limit number of files in one dir
              . Server will only try to clean the dir it caches to
              . Added -C (--clean-cache) option to gorg to clean whole cache tree with a cron job e.g.
2005-01-10  Add filter functionality:
              . Added -F (--filter) option to process STDIN
                Both gorg and gorg.cgi can be used as a filter.
                gorg.cgi needs the -F (--filter) option on its command line
                gorg will behave as a filter when no option are used *and* data is piped into it
              . Added -W (--web) Starting the web server is the default on the cmd line
                but it might need to be told when not started interactively

======================================================
           
2005-01-10 gorg-0.5 initiated

2005-04-04  . Trivial code cleanup to get rid of signal catching that was introduced
              before issues with mod_fastcgi and ruby-fcgi had been pinpointed.
            . wasCash can be set to 0 to disable cache cleaning, use `gorg --clean-cache` to clean up
            . xsl.c includes trivial patch that was applied the the -0.4.ebuild for 64-bit systems
            . ebuild now knows apache and fastcgi USE flags to depend on apache and fastcgi (mod-fcgid) 

======================================================
           
2005-07-04 gorg-0.6 initiated

2005-07-04  . Pass parameter with path name of requested file (e.g. $link='/doc/en/index.xml')
            . Fix segfault when using a string stylesheet instead of a file
              (undocumented free in libxslt lead to a double free)
            . Issue http header with  text/html, text/xml or application/xhtml+xml as content-type,
              replace application/xhtml+xml with text/html if browser does not accept it
            . Stand-alone web server (ie. webrick) fails (500) like the (f)cgi versions when lib(xslt|xml2) return a warning
              It's just too confusing to have pages that work with "gorg &" generate errors once live with apache or lighttpd
            . Make stand-alone web server use index automatically when requesting a dir and index.xml exists
            . Add mount points for webrick (eg. for /cgi-bin or /images) to be handled by stock FileHandler (ie. not gorg)
            . xsl extension returns messages output with xsl:message that begin with '%%GORG%%'
            . Make xsl extension return requested remote file names (ftp:// & http://) and let
              caller decide how to handle caching of those resources
            . gorg has decided not to cache objects that request foreign URIs
            . Allow cookies to be passed to transforms and transforms to set cookies with
              xsl:message '%%GORG%%Set-Cookie(cookiename)key=value'
              eg.  <xsl:message>
                    <xsl:value-of select="concat('%%GORG%%Set-Cookie(prefs)SL=',$searchLang)"/>
                   <xsl:message>
              NOTE: Setting several cookies will not work properly in some cases with the stand-alone web server
                    http://rubyforge.org/tracker/?func=detail&aid=2199&group_id=426&atid=1698
              *Update* My patch has been accepted upstream and should make it into an upcoming ruby release
            . Allow http redirect (301) with xsl:message '%%GORG%%Redirect=newURI'
            . Add charset= to http header for xml files what have encoding="bleh"
            . Deprecate zipCache param in favour of zipLevel (0-9) and add support for mod_gzip
              zipLevel makes gorg compress cached data *and* returned data if client requests it (accept_encoding=gzip)
              Decompressing compressed cached data to have it recompressed on output by apache with mod_deflate is a waste of CPU
              Besides, gorg can now be used with web servers that do not support deflating dynamic data (eg. lighttpd)
              In short, 0   == gorg does not compress data, neither in its cache nor its output
                               Of course, you may use mod_deflate or similar to make your web server support gzip encoding

lib/gorg/base.rb
###   Copyright 2004,   Xavier Neys   (neysx@gentoo.org)
# #
# #   This file is part of gorg.
# #
# #   gorg is free software; you can redistribute it and/or modify
# #   it under the terms of the GNU General Public License as published by
# #   the Free Software Foundation; either version 2 of the License, or
# #   (at your option) any later version.
# #
# #   gorg is distributed in the hope that it will be useful,
# #   but WITHOUT ANY WARRANTY; without even the implied warranty of
# #   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# #   GNU General Public License for more details.
# #
# #   You should have received a copy of the GNU General Public License
# #   along with Foobar; if not, write to the Free Software
###   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

module Gorg
  Version = "0.6"
end
  
# Some required stuff for gorg
require 'time'

require 'gorg/xsl'
require 'gorg/log'
require 'gorg/cache'
require 'timeout'
require 'cgi'
require 'stringio'
require 'zlib'
require 'ipaddr'


module Gorg

  def xproc(path, params, list=false, printredirect=false)
    # Process file through xslt passing params to the processor
    # path should be the absolute path of the file, i.e. not relative to DocumentRoot
    #
    #    Since 0.4, path can also be a string containing
    #    the actual xml to be processed
    #
    # Use default stylesheet if none can be found in the file
    # Return a list of files read by the processor (useful to do caching) if requested
    #
    # Return an error condition and, hopefully, some useful output
    # Do not raise any exception
    # In most cases, an error will result in no output but
    # the xslt processor can consider some errors as warnings and
    # return the best result it could come up with along with a warning
    # e.g. if a file used in a document() function cannot be found,
    # the xslt processor will return some output and a warning.
    # It's up to the caller to decide whether to use the output or b0rk
    #
    # The return value is an array of 2 to 4 items: [{}, "", [[]], []]
    # 1. hash with error information, its keys are
    # 1.a  "xmlErrCode"  0 is no error, -9999 means an exception has been raised in this block (unlikely),
    #      anything else is an error code (see /usr/include/libxml2/libxml/xmlerror.h)
    # 1.b  "xmlErrLevel" again, from libxml2, 0==OK, 1==Warning, 2==Error, 3==Fatal
    # 1.c  "xmlErrLevel" again, from libxml2, some explanation about what went wrong
    # 2. output from xsltprocessor (or error message from a raised exception)
    # 3. list of files that the xslt processor accessed if the list was requested,
    #    paths are absolute, i.e. not relative to your docroot.
    #    Each entry is an array [access type, path] with access_type being
    #      "r" for read, "w" for written (with exsl:document) or "o" for other (ftp:// or http://)
    # 4. array of CGI::Cookie to be sent back
    #
    # Examples: [{"xmlErrMsg"=>"blah warning blah", "xmlErrCode"=>1509, "xmlErrLevel"=>1}, "This is the best XSLT could do!", nil]
    #           [{"xmlErrCode"=>0}, "Result of XSLT processing. Well done!", ["/etc/xml/catalog","/var/www/localhost/htdocs/doc/en/index.xml","/var/www/localhost/htdocs/dtd/guide.dtd"]]

    xsltproc = Gorg::XSL.new
    xsltproc.xroot = $Config["root"]
    # Grab strings from xsl:message
    xslMessages = []
    # Does the caller want a list of accessed files?
    xsltproc.xtrack = list; filelist = Array.new
    # Process .xml file with stylesheet(s) specified in file, or with default stylesheet
    xsltproc.xml = path
    # Look for stylesheet href (there can be more than one)
    regexp = Regexp.new('<\?xml-stylesheet.*href="([^"]*)".*')
    l = $Config["headXSL"] ; styles = Array.new
    if FileTest.file?(path) then
      # Path is indeed a file name
      IO.foreach(path) { |line|
        styles << $1 if regexp.match(line)
        break if (l-=1) == 0
      }
    else
      # Scan xml for stylesheet names
      path.each_line { |line| styles << $1 if regexp.match(line) }
    end
    # Use default stylesheet if none were found in the doc
    styles << $Config["defaultXSL"] if styles.length == 0
    # Add params, we expect a hash of {param name => param value,...}
    xsltproc.xparams = params
    # Process through list of stylesheets
    firstErr = {}
    while xsltproc.xsl = styles.shift
      xsltproc.process
      filelist += xsltproc.xfiles if xsltproc.xtrack?
      # Break and raise 301 on redirects
      xsltproc.xmsg.each { |r|
        if r =~ /Redirect=(.+)/ then
          if printredirect then
            STDERR.puts "Location: #{$1}"
          else
            raise Gorg::Status::MovedPermanently.new($1)
          end
        end
      }
      xslMessages += xsltproc.xmsg
      # Remember 1st warning / error
      firstErr = xsltproc.xerr if firstErr["xmlErrLevel"].nil? && xsltproc.xerr["xmlErrLevel"] > 0
      # B0rk on error, an exception should have been raised by the lib, but, er, well, you never know
      break if xsltproc.xerr["xmlErrLevel"] > 1 
      xsltproc.xml = xsltproc.xres
    end
    # Keep 1st warning / error if there has been one
    firstErr = xsltproc.xerr if firstErr["xmlErrLevel"].nil?
    # Return values
    [ firstErr, xsltproc.xres, (filelist.uniq if xsltproc.xtrack?), xslMessages ]
  rescue => ex
    if ex.respond_to?(:errCode) then
      # One of ours (Gorg::Status::HTTPStatus)
      # Propagate exception
      raise
    else
      debug "in xproc exception handler: #{ex.inspect} // #{xsltproc.xerr.inspect}"
      # Return exception message and an error hash as expected from the xslt processor
      # Use error codes that the xslt lib might have returned
      [ if (xsltproc.xerr["xmlErrCode"]||-1) == 0 then
        { "xmlErrMsg"   => ex.to_s,
          "xmlErrCode"  => 9999,
          "xmlErrLevel" => 3
        }
        else
        { "xmlErrMsg"   => xsltproc.xerr["xmlErrMsg"] || ex.to_s,
          "xmlErrCode"  => xsltproc.xerr["xmlErrCode"],
          "xmlErrLevel" => xsltproc.xerr["xmlErrLevel"]
        }
        end ,
        ex.to_s,
        (filelist.uniq if xsltproc.xtrack?)
      ]
    end
  end
  
  # HTTP status codes and html output  
  module Status
    class HTTPStatus < StandardError
      def html(err="")
        <<-EOR
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<HTML>
<HEAD><TITLE>#{errSts}</TITLE></HEAD>
<BODY>
<H1>#{errLabel}</H1>
<font color="#FF0000">#{err}</font>
<HR>
</BODY>
</HTML>
      EOR
      end
      def errSts
        "#{errCode} #{errLabel}"
      end
      # Default is unknown error
      def errLabel
        "Undefined Error"
      end
      def errCode
        999
      end
      def header
        {'Status' => errSts}
      end
    end
    
    class NotModified < HTTPStatus
      def initialize(stat)
        # 304 needs to send ETag and Last-Modified back
        @mstat=stat
      end
      def header
        {'Last-Modified' => @mstat.mtime.httpdate.dup, 'ETag' => makeETag(@mstat).dup}.merge(super)
      end
      def html
        ""
      end
      def errLabel
        "Not Modified"
      end
      def errCode
        304
      end
    end
    
    class MovedPermanently   < HTTPStatus
      def initialize(loc)
        # 301 needs to send Location:
        @location=loc
      end
      def errLabel
        "Moved Permanently"
      end
      def errCode
        301
      end
      def header
        {'Location' => @location}.merge(super)
      end
      def html
        # RFC says "should" not "must" add a body
        ""
      end
      def html301 # Not used
        <<-EO301
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>301 Moved Permanently</title>
</head><body>
<h1>Moved Permanently</h1>
<p>The document has moved <a href="#{@location}">here</a>.</p>
</body></html>
        EO301
      end
    end
    
    class Forbidden   < HTTPStatus
      def errLabel
        "Forbidden"
      end
      def errCode
        403
      end
    end
    
    class NotFound    < HTTPStatus 
      def errLabel
        "Not Found"
      end
      def errCode
        404
      end
    end
    
    class NotAllowed    < HTTPStatus 
      def errLabel
        "Method Not Allowed"
      end
      def header
        {'Allow'=>'GET,HEAD'}.merge(super)
      end
      def errCode
        405
      end
    end
    
    class SysError    < HTTPStatus 
      def errLabel
        "Internal Server Error"
      end
      def errCode
        500
      end
    end
  end #Status module
  
  
  def gorgInit
    # Initialize gorg, i.e. read config file, init cache, ...
    # Simply build a hash of params => value in a global variable called $Config
    
    # Set up default values
    $Config = { "AppName" => "gorg",    # Used for syslog entries, please keep 'gorg' (cannot be changed in config file)
                "root" => nil,          # No root dir by default (cgi uses DOCUMENT_ROOT from its environment)
                "port" => 8000,         # Used for stand-alone web server (WEBrick)
                "headXSL" => 12,        # Only read 12 lines in xml files to identify required stylesheets
                "defaultXSL" => nil,    # No default stylesheet, how could I guess?
                "cacheDir" => nil,      # No cache by default. Directory must exist and be writable.
                "cacheTTL" => 0,        # Number of seconds after which a document is considered too old, 0=never
                "cacheSize" => 40,      # in MegaBytes, max size of cache, used when autocleanig
                "zipLevel" => 2,        # Compresion level used for gzip support (HTTP accept_encoding) (0-9, 0=none, 9=max)
                "maxFiles" => 9999,     # Max number of files in a single directory in the cache tree
                "cacheTree" => 0,       # Use same tree as on site in cache, 0 = disabled
                "cacheWash" => 0,       # Clean cache automatically and regularly when a store into the cache occurs. 0 = disabled
                                        #  gorg cleans up if random(param_value) < 10. It will only clean same dir it caches to, not whole tree.
                                        # i.e. a value<=10 means at every call (not a good idea), 100 means once/10 stores, 1000 means once/100 stores
                "logLevel" => 4,        # INFO, be slightly verbose by default (messages go to syslog) OFF, FATAL, ERROR, WARN, INFO, DEBUG = 0, 1, 2, 3, 4, 5
                "passthru" => true,     # Allow return of requested file without processing it if passthru="anything but 0" is passed
                "acceptCookies" =>false,# Allow cookies in & out of transforms
                "linkParam" => "link",  # Pass pathname of requested file in 'link' param to xsl transform
                "HTTP_HOST" => nil,     # Pass host value from HTTP header to xsl transform
                "accessLog" => "syslog",# or a filename or STDERR, used to report hits from WEBrick, not used by cgi's
                "autoKill" => 0,        # Only used by fastCGI, exit after so many requests (0 means no, <=1000 means 1000). Just in case you fear memory leaks.
                "in/out" => [],         # (In/Ex)clude files from indexing
                "mounts" => [],         # Extran mounts for stand-alone server
                "listen" => "127.0.0.1" # Let webrick listen on given IP
            }
    # Always open syslog
    @syslog = Gorg::Log::MySyslog.new($Config["AppName"])
    $Log = Gorg::Log::MyLog.new(@syslog, 5) # Start with max
    
    # Check for config file
    configf = ENV["GORG_CONF"]||"/etc/gorg/gorg.conf"
    raise "Cannot find config file (#{configf})" unless FileTest.file?(configf) and FileTest.readable?(configf)
    file = IO.read(configf)
    parseConfig($Config, file)

    # Init cache
    Cache.init($Config) if $Config["cacheDir"]
    
    # Set requested log level
    $Log.level = $Config["logLevel"]
  rescue
    error("Gorg::init failed: #{$!}")
    STDERR.puts("Gorg::init failed: #{$!}")
    exit(1)
  end

  def scanParams(argv)
    # Scan argv for --param paramName paramValue sequences
    # params are removed from argv
    # Return a hash of {"name" => "value"}
    h = Hash.new
    while idx = argv.index('--param')
      break if argv.length <= idx+2   # We need at least 2 more args after --param
      argv.delete_at(idx)             # Remove --param from argv
      name  = argv.delete_at(idx)     # Remove param name from argv
      value = argv.delete_at(idx)     # Remove param value from argv
      h[name] = value                 # Add entry in result
    end
    
    h if h.length > 0
  end  
  
  private
  def parseConfig(h, config)
    config.each_line {|line|
      line.strip!
      next if line.length == 0 or line[0,1] == '#' # Skip blank lines and comments
      raise "Invalid Configuration (#{line})" unless line =~ /^([a-zA-Z_]*)\s*=\s*/
      param = $1
      value = $'
      # If value starts with ' or ", it ends with a similar sign and does not accept any in the value, no escaping... We keep it simple
      # otherwise, it ends with EOL or first space
      if value =~ /["'](.*)['"]/ then
        value = $1
      end
      value.strip!
      raise "No value for #{param}" unless value.length > 0
      # Check param / value (only syntactical checks here)
      case param.downcase
      when "root"
       h["root"] = value
      when "port"
       h["port"] = value.to_i
      when "passthru"
       h["passthru"] = value.squeeze != "0"
      when "acceptcookies"
       h["acceptCookies"] = value.squeeze == "1"
      when "linkparam"
       if value =~ /^\s*([a-zA-Z]+)\s*$/ then
         h["linkParam"] = $1
       else
         h["linkParam"] = nil
       end
      when "httphost"
        hosts = value.squeeze(" ")
        case hosts
          when /^0?$/ 
            hh = nil
          when "*"
            hh = ["*"]
          else
            hh = hosts.split(" ")
            # Add IPs
            hosts.split(" ").each { |ho|
              begin
                hh += TCPSocket.gethostbyname(ho)[3..-1] if ho != '*'
              rescue
                # Ignore
                nil
              end
            }
            hh.uniq!
        end
        h["httphost"] = hh
      when "headxsl"
       h["headXSL"] = value.to_i
      when "defaultxsl"
       h["defaultXSL"] = value
      when "cachedir"
       h["cacheDir"] = value
      when "cachettl"
       h["cacheTTL"] = value.to_i
      when "cachesize"
       h["cacheSize"] = value.to_i
      when "maxfiles"
       h["maxFiles"] = value.to_i
      when "cachetree"
       h["cacheTree"] = value.squeeze != "0"
      when "ziplevel"
       if value =~ /^\s*([0-9])\s*$/ then
         h["zipLevel"] = $1.to_i
       else
         h["zipLevel"] = 2
       end
      when "cachewash"
       h["cacheWash"] = value.to_i
      when "loglevel"
       h["logLevel"] = value.to_i
      when "accesslog"
       h["accessLog"] = value
      when "autokill"
       h["autoKill"] = value.to_i
      when "listen"
       begin
         ip = IPAddr.new(value)
         h["listen"] = ip.to_s
       rescue
         h["listen"] = "127.0.0.1"
       end
      when "dbconnect"
        h["dbConnect"] = value
      when "dbuser"
        h["dbUser"] = value
      when "dbpassword"
        h["dbPassword"] = value
      when "exclude"
        h["in/out"] << [false, Regexp.new(value)]
      when "include"
        h["in/out"] << [true,  Regexp.new(value)]
      when "fpath_to_lang"
        h["flang"] = Regexp.new(value)
      when "xpath_to_lang"
        h["xlang"] = value      
      when "mount"
        if value =~ /^([^\s]+)\s+ON\s+(.+)$/i then
          h["mounts"] << [$1, $2]
        end
      else
        raise "Unknown parameter (#{param})"
      end
    }
  rescue
    raise "Could not parse config file: #{$!}"
  end
  
  # Utilities
  def contentType(aMsg)
    # Find the Content-Type=xxx/yyy line in aMsg
    # from the Meta file in the cache
    ct = nil
    aMsg.each { |s|
      if s =~ /^Content-Type:(.+)$/ then
        ct = $1
        break
      end
    }
    ct
  end
  
  def setContentType(data)
    # Set content-type according to x(ht)ml headers
    charset = nil
    if data =~ /^<\?xml .*encoding=['"](.+)['"]/i then
      charset = $1 if $1
      # XML / XHTML
      if data[0..250] =~ /^<\!DOCTYPE\s+html/i then
        # XHTML
        ct = 'application/xhtml+xml'
      else
        # XML
        ct = 'text/xml'
      end
      if charset then
        ct << "; charset=#{charset}"
      end
    elsif data =~ /^<\!DOCTYPE\s+html\sPUBLIC\s(.+DTD XHTML)?/i then
      # (X)HTML
      if $1 then
        # XHTML
        ct = 'application/xhtml+xml'
      else
        # HTML
        ct = 'text/html'
      end
    elsif data =~ /<html/i then
      # HTML
      ct = 'text/html'
    else
      # TXT
      ct = 'text/plain'
    end
    ct
  end

  def makeCookies(aMsg)
    # Make an array of CGI::Cookie objects
    # msg is expected to be an array of strings like 'Set-Cookie(name)value=param'
    # (output by the xsl transform with xsl:message)
    cookies = Hash.new
    aMsg.each { |s|
      if s =~ /^Set-Cookie\(([^\)]+)\)([a-zA-Z0-9_-]+)=(.+)$/ then
        # $1 = cookie name   $2 = key   $3 = value
        if cookies.has_key?($1) then
          cookies[$1] << "#{$2}=#{$3}"
        else
          cookies[$1] = ["#{$2}=#{$3}"]
        end
      end
    }
    if cookies.length > 0 then
      # Make CGI::Cookie objects
      cookies.map { |k,v| 
        CGI::Cookie.new('name' => k, 'value' => v, 'expires' => Time.now + 3600*24*30)
      }
    else
      nil
    end
  end
  
  def cookies_to_params(cookies)
    # Turn array of CGI::Cookie objects into a Hash of key=>value
    # cookies is a hash, forget the keys,
    # each value should be an array of strings, each string should be like 'param=value'
    h = {}
    cookies.values.each { |v|
      if v.class==String and v =~ /^([a-zA-Z0-9_-]+)=(.+)$/ then
        h[$1] = $2
      elsif v.class==Array then
        v.each { |vv|
          if vv.class==String and vv =~ /^([a-zA-Z0-9_-]+)=(.+)$/ then
            h[$1] = $2
          end
        }
      elsif v.class==CGI::Cookie then
        v.value.each { |vv|
          if vv.class==String and vv =~ /^([a-zA-Z0-9_-]+)=(.+)$/ then
            h[$1] = $2
          end
        }
      end
    }
    h
  rescue
    error "Could not parse cookies (#{$!}) "
    {}
  end
  
  def notModified?(fstat, etags, ifmodsince)
    # Decide whether file has been modified according to either etag, last mod timestamp or both
    # If both If-None-Match and If-Modified-Since request header fields are present,
    # they have to be tested both
    res = false
    if fstat then
      a = etags.to_a
      if ifmodsince && etags then
        res = (ifmodsince >= fstat.mtime) && (a.include?(makeETag(fstat)) || a.include?('*'))
      elsif etags
        res = a.include?(makeETag(fstat)) || a.include?('*')
      elsif ifmodsince
        res = ifmodsince >= fstat.mtime
      end
    end
    # Return result
    res
  end  
  
  def split_header_etags(str)
    # Split header values expected as "value1", "value2", ... into an array of strings
    str.scan(/((?:"(?:\\.|[^"])+?"|[^",]+)+)(?:,\s*|\Z)/xn).collect{|v| v[0].strip }
  end
 
  def makeETag(st)
    # Format file stat object into an ETag using its size & mtime
    # Parameter can either be a filename or a stat object
    st = File.stat(st) unless st.respond_to?(:ino)
    sprintf('"%x-%x"', st.size, st.mtime.to_i)
  end
  
  def gzip(data, level)
    gz = ""
    io = StringIO.new(gz)
    gzw = Zlib::GzipWriter.new(io, level)
    gzw.write data
    gzw.close
    gz
  end
  
  def gunzip(data)
    io = StringIO.new(data)
    gzw = Zlib::GzipReader.new(io)
    gunz = gzw.read
    gzw.close
    gunz
  end

end
